core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java

```diff
@@ -260,17 +260,25 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
                     final long maxContentLength = decodedReq.maxRequestLength();
                     final long transferredLength = decodedReq.transferredBytes();
                     if (maxContentLength > 0 && transferredLength > maxContentLength) {
-                        final ContentTooLargeException cause =
-                                ContentTooLargeException.builder()
-                                                        .maxContentLength(maxContentLength)
-                                                        .contentLength(req.headers())
-                                                        .transferred(transferredLength)
-                                                        .build();
-                        fail(id, decodedReq.headers(), HttpStatus.REQUEST_ENTITY_TOO_LARGE,
-                             Http2Error.CANCEL, null, cause);
-                        // Wrap the cause with the returned status to let LoggingService correctly log the
-                        // status.
-                        decodedReq.close(HttpStatusException.of(HttpStatus.REQUEST_ENTITY_TOO_LARGE, cause));
+                        final Routed<ServiceConfig> routed = req.route();
+                        if (routed != null && routed.route().isFallback()) {
+                            // Don't need to return an error response. `FallbackService` immediately respond to
+                            // the request without consuming data.
+                            this.req = null;
```

Comment:
Question) 

1) Is there no need to also close the request in case there are elements in the queue that need cleaning up? (e.g. callbacks waiting on a future that is never completed)
2) Is there no need to set `discarding = true` so that incoming `HttpMessage`s are ignored correctly?


```suggestion
                           discarding = true;
                            this.req.close();
                            this.req = null;
```

Reply:
> Is there no need to also close the request in case there are elements in the queue that need cleaning up? (e.g. callbacks waiting on a future that is never completed)

I thought the request is cleaned up when a response is successfully sent by `FallbackService`. However, if there is a decorator that aggregates the request, the future won't be completed until a timeout is triggered. For that case, `req.close()` makes sense.

> Is there no need to set discarding = true so that incoming HttpMessages are ignored correctly?

If a non-`HttpRequest` type object is received when `req` is null, the connection will be reset immediately. I hesitated to set `discarding` to true because `FallbackService` does not send `connection: close` to the client. Speaking of `connection: close`, it would be worth sending `connection: close` when a request content exceeds the limitation.

Reply:
> If a non-HttpRequest type object is received when req is null, the connection will be reset immediately

I encountered the following warning log with the test snippet. It seems like after the response is sent, another 400 is sent:
https://github.com/line/armeria/blob/f4781d52a34f874e9459994afd8de763c9608480/core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java#L236
```
22:17:45.969 [armeria-common-worker-nio-2-2] WARN  c.l.a.client.Http1ResponseDecoder - Unexpected exception:
com.linecorp.armeria.common.ProtocolViolationException: unexpected message type: io.netty.handler.codec.http.DefaultLastHttpContent (expected: io.netty.handler.codec.http.HttpResponse, channel: [id: 0xa01f997d, L:/127.0.0.1:51010 ! R:127.0.0.1/127.0.0.1:51006], resId: 2)
	at com.linecorp.armeria.client.Http1ResponseDecoder.failWithUnexpectedMessageType(Http1ResponseDecoder.java:273)
```

Can you check the following snippet?

```
@RegisterExtension
static ServerExtension server = new ServerExtension() {
    @Override
    protected void configure(ServerBuilder sb) throws Exception {
        sb.requestTimeoutMillis(Long.MAX_VALUE);
        sb.idleTimeoutMillis(Long.MAX_VALUE);

        sb.maxRequestLength(100);
        sb.decorator((delegate, ctx, req) -> HttpResponse.delayed(HttpResponse.of(200), Duration.ofSeconds(3)));
        sb.decorator(LoggingService.builder()
                                   .successfulResponseLogLevel(LogLevel.DEBUG)
                                   .newDecorator());
        sb.service("/", (ctx, req) -> HttpResponse.of(200));
    }
};

@Test
void testMaxContentLength() {
    final HttpRequestWriter streaming = HttpRequest.streaming(HttpMethod.POST, "/not-exist");
    final HttpResponse response = WebClient.builder(SessionProtocol.H1C, server.endpoint(SessionProtocol.H1C))
                                           .writeTimeoutMillis(Long.MAX_VALUE)
                                           .responseTimeoutMillis(Long.MAX_VALUE)
                                           .build()
                                           .execute(streaming);
    for (int i = 0; i < 10; i++) {
        streaming.write(HttpData.ofUtf8(Strings.repeat("a", 30)));
    }
    streaming.close();
    System.out.println(response.aggregate().join());
}
```

Reply:
> Can you check the following snippet?

Checked in the tests and fixed Http{1,2}RequestDecoder to pass the tests.